R-packages/covidcast/R/plot.R

```diff
@@ -582,69 +582,130 @@ plot_bubble = function(x, time_value = NULL, include = c(), range = NULL,
+      map_df = map_df %>% 
+        dplyr::filter(.$STUSPS %in% include)
+    }
   }
 
-  # Determine which locations are missing
-  map_mis = rep(1, length(map_geo))
-  map_mis[map_geo %in% geo] = 0
-
+  # Important: make into a factor and set the levels (for the legend)
+  # Factor bubble values before splitting up into mainland and non-main layers
+  map_df$bubble_val <- factor(map_df$bubble_val, levels = breaks)
+  
+  # Explicitly drop zeros (and from levels) unless were asked not to
+  if (!isFALSE(params$remove_zero)) {
+    map_df$bubble_val[map_df$bubble_val == 0] = NA
+    levels(map_df$bubble_val)[levels(map_df$bubble_val) == 0] = NA
+  }
+  
   # Warn if there's any missing locations
-  if (sum(map_mis == 1) > 0) {
+  if (sum(map_df$back_color == missing_col) > 0) {
     warning("Bubble maps can be hard to read when there is missing data;",
             "the locations without data are filled in gray.")
   }
+  
+  # Adjust map data for non mainland
+  main_df = shift_main(map_df)
+  hawaii_df = shift_hawaii(map_df)
+  alaska_df = shift_alaska(map_df)
+  pr_df = shift_pr(map_df)
 
-  # Create the polygon layer
+  # Create the polygon layers
   aes = ggplot2::aes
   geom_args = list()
   geom_args$color = border_col
   geom_args$size = border_size
-  geom_args$fill = c("white", missing_col)[map_mis + 1]
-  geom_args$mapping = aes(x = x, y = y, group = group)
-  geom_args$data = map_df
-  polygon_layer = do.call(ggplot2::geom_polygon, geom_args)
-
-  # Retrieve coordinates for mapping
-  # Reading from usmap files to ensure consistency with borders
-  if (attributes(x)$metadata$geo_type == "county") {
-    centroids = covidcast::county_geo[covidcast::county_geo$fips %in% map_geo, ]
-    cur_geo = centroids$fips
-    cur_val = rep(NA, length(cur_geo))
-  }
-  else if (attributes(x)$metadata$geo_type == "state") {
-    centroids = covidcast::state_geo
-    centroids$abbr = tolower(centroids$abbr)
-    centroids = centroids[centroids$abbr %in% map_geo, ]
-    cur_geo = centroids$abbr
-    cur_val = rep(NA, length(cur_geo))
-  }
-
-  # Overwrite the values for observed locations
-  cur_obs = cur_geo[cur_geo %in% geo]
-  cur_val[cur_geo %in% geo] = dis_fun(val[cur_obs])
-
-  # Important: make into a factor and set the levels (for the legend)
-  cur_val = factor(cur_val, levels = breaks)
+  geom_args$mapping = aes(geometry=geometry)
+  geom_args$fill = main_df$back_color
+  geom_args$data = main_df
+  main_layer = do.call(ggplot2::geom_sf, geom_args)
+  geom_args$fill = pr_df$back_color
+  geom_args$data = pr_df
+  pr_layer = do.call(ggplot2::geom_sf, geom_args)
+  geom_args$fill = hawaii_df$back_color
+  geom_args$data = hawaii_df
+  hawaii_layer = do.call(ggplot2::geom_sf, geom_args)
+  geom_args$fill = alaska_df$back_color
+  geom_args$data = alaska_df
+  alaska_layer = do.call(ggplot2::geom_sf, geom_args)
+  
+  # Change geometry to centroids
+  # Use centroid coordinates for plotting bubbles
+  suppressWarnings({
```

Comment:
Why is `suppressWarnings` needed here? Does `st_centroid` raise warnings? If it does, and we need to ignore them, we should have a comment saying what the warning is and why it's safe to ignore

Reply:
Is this sufficient?  This knowledge is based on stack overflow answers and a visual inspection of the bubbles.

Reply:
I think the new comment is sufficient, yeah. Even if the warning is right and the centroids aren't exact, we don't really need them to be perfectly precise